### PR TITLE
[Merchants] Add Merchant Data Bucket capability.

### PR DESCRIPTION
**Pull request by Kinglykrab**
- Allows server operators to limit merchant items based on data bucket values and comparisons.
- Adds 3 columns, bucket_name, bucket_value, and bucket_comparison to merchantlist table.
- Bucket is checked based on GetBucketKey()-bucket_name.
- Buckets are mass loaded when using the merchant so it's not a database call per item, just a grouping of all their buckets from the start.
- This is a nearly year old pull request redone for master.
- bucket_comparison Values are as follows:
```
0: bucket_name == bucket_value
1: bucket_name != bucket_value
2: bucket_name >= bucket_value
3: bucket_name <= bucket_value
4: bucket_name > bucket_value
5: bucket_name < bucket_value
6: bucket_name is any of pipe(|)-separated bucket_value
7: bucket_name is not any of pipe(|)-separated bucket_value
8: bucket_name is between first and second value of pipe(|)-separated bucket_value
9: bucket_name is not between first and second value of pipe(|)-separated bucket_value
```

Example logs when I was testing.
image

### CODE REVIEW

File: zone/zone.cpp [hunk 1]
```diff
@@ -634,25 +621,28 @@ void Zone::LoadNewMerchantData(uint32 merchantid) {
         return;
 	}
 
-	for (auto row = results.begin(); row != results.end(); ++row) {
+	for (auto row : results) {
```
Comment (Akkadius) [inline zone/zone.cpp:624]: `&row` to avoid unnecessary copy

File: zone/zone.cpp [hunk 2]
```diff
@@ -605,26 +605,13 @@ void Zone::LoadTempMerchantData()
 
 void Zone::LoadNewMerchantData(uint32 merchantid) {
 
-	std::list<MerchantList> merlist;
+	std::list<MerchantList> merchant_list;
 
-	std::string query = fmt::format(
-		SQL(
```
Comment (Akkadius) [inline zone/zone.cpp:611]: Any reason we scrunched up the query (readability) ? Repository usage would also clean this up

File: zone/zone.cpp (hunk 1 above)
Comment (Akkadius) [reply zone/zone.cpp:624]: And const

File: zone/client_process.cpp [hunk 3]
```diff
@@ -813,152 +813,168 @@ void Client::BulkSendInventoryItems()
...
-	std::list<TempMerchantList> tmp_merlist = zone->tmpmerchanttable[npcid];
-	std::list<TempMerchantList>::iterator tmp_itr;
 
-	uint32 i = 1;
-	uint8 handychance = 0;
-	for (itr = merlist.begin(); itr != merlist.end() && i <= numItemSlots; ++itr) {
-		MerchantList ml = *itr;
-		if (ml.probability != 100 && zone->random.Int(1, 100) > ml.probability)
-			continue;
+	auto client_data_buckets = GetMerchantDataBuckets();
```
Comment (Akkadius) [inline zone/client_process.cpp:834]: Thanks for turning this into a single DB call from the last time

**Review by Akkadius — APPROVED** [review]
LGTM.

Really make sure there's docs around this so users know how to find it (eqemu.docs.io) and more importantly know how to use it.

File: zone/zone.cpp (hunk 2 above)
Comment (Kinglykrab) [reply zone/zone.cpp:611]: I can put it back the way it was, I figured it was more readable this way.

**Comment by Kinglykrab** [comment]
Documented and available already: https://docs.eqemu.io/server/scripting/merchant-data-buckets/

Going to wait on people to put some eyes back on this so if there's something more I need to do here, I can before merge.

File: zone/zone.cpp (hunk 1 above)
Comment (Akkadius) [reply zone/zone.cpp:636]: Yeah the underlying doesn't exactly play nicely

**Comment by Akkadius** [comment]
Docs look good

**Pull request merged**